### PR TITLE
Implement `ViewOp::evaluate`.

### PR DESCRIPTION
**Pull request by wujingyue**

### CODE REVIEW

**Comment by wujingyue** [comment]
!build

File: csrc/ir/nodes.cpp [hunk 1]
```diff
@@ -2328,6 +2328,28 @@ std::string ViewOp::toInlineString(int indent_size) const {
   NVF_CHECK(false, "Tensor op can not be printed inline");
 }
 
+std::vector<PolymorphicValue> ViewOp::evaluate(
+    const ExpressionEvaluator& original_ee,
+    const std::vector<PolymorphicValue>& inputs) const {
+  NVF_ERROR(inputs.size() == 1);
+  at::Tensor in_tensor = inputs[0].as<at::Tensor>();
```
Comment (zasdfgbnm) [inline csrc/ir/nodes.cpp:2335]: nit:
```suggestion
  const at::Tensor& in_tensor = inputs[0].as<at::Tensor>();
```

File: test/test_evaluator.cpp [hunk 2]
```diff
@@ -558,4 +558,72 @@ TEST_F(ExprEvalTest, Permute) {
...
+  EXPECT_THAT(out_tensor.sizes(), ElementsAre(6, 9));
+  EXPECT_THAT(out_tensor.strides(), ElementsAre(1, 8));
+}
+
+TEST_F(ExprEvalTest, Reshape_ForwardBroadcast) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  TensorView* in =
+      TensorViewBuilder().shape({-1, 6}).dtype(DataType::Float).build();
```
Comment (zasdfgbnm) [inline test/test_evaluator.cpp:590]: Should we set the first dim to expanded?
https://github.com/NVIDIA/Fuser/blob/5a22d44ee8a4d8a8a6dd2ce63016552be61fbc43/csrc/ir/interface_nodes.h#L630

File: test/test_evaluator.cpp [hunk 3]
```diff
@@ -558,4 +558,72 @@ TEST_F(ExprEvalTest, Permute) {
...
+
+TEST_F(ExprEvalTest, Reshape_SplitBroadcast) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  TensorView* in =
+      TensorViewBuilder().shape({-1, 6}).dtype(DataType::Float).build();
+  fusion.addInput(in);
+  TensorView* out = reshape(
+      in, {size(in, 0), IrBuilder::create<Val>(2), IrBuilder::create<Val>(3)});
```
Comment (zasdfgbnm) [inline test/test_evaluator.cpp:615]: Same here: https://github.com/NVIDIA/Fuser/blob/5a22d44ee8a4d8a8a6dd2ce63016552be61fbc43/csrc/ir/interface_nodes.h#L630

File: test/test_evaluator.cpp (hunk 2 above)
Comment (wujingyue) [reply test/test_evaluator.cpp]: TIL! This made it possible for me to test expanded broadcasts better and actually find a bug.

File: test/test_evaluator.cpp (hunk 3 above)
Comment (wujingyue) [reply test/test_evaluator.cpp:621]: Done.

File: csrc/ir/nodes.cpp [hunk 4]
```diff
@@ -2328,6 +2328,32 @@ std::string ViewOp::toInlineString(int indent_size) const {
...
+  out_shape.reserve(out_rfactor.size());
+  for (IterDomain* id : out_rfactor) {
+    out_shape.push_back(
+        ee.evaluate(id->getMaybeExpandedExtent()).as<int64_t>());
+  }
+
+  // TODO: check allocation domain and contiguity.
+
+  // Use `at::Tensor::reshape` instead of `at::Tensor::view` because `ViewOp`
+  // doesn't always produce an alias. For example, when merging an expanded
```
Comment (zasdfgbnm) [inline csrc/ir/nodes.cpp:2351]: I think the problem is, `torch.view` always produce an alias, but our ViewOp is wider than that.

**Review by zasdfgbnm — APPROVED** [review]

File: csrc/ir/nodes.cpp [hunk 5]
```diff
@@ -2328,6 +2328,32 @@ std::string ViewOp::toInlineString(int indent_size) const {
...
 }
 
+std::vector<PolymorphicValue> ViewOp::evaluate(
+    const ExpressionEvaluator& original_ee,
+    const std::vector<PolymorphicValue>& inputs) const {
+  NVF_ERROR(inputs.size() == 1);
+  const at::Tensor& in_tensor = inputs[0].as<at::Tensor>();
+
+  ExpressionEvaluator ee(original_ee);
+  ee.bind(in(), in_tensor);
```
Comment (Priya2698) [inline csrc/ir/nodes.cpp:2338]: Question: Why can't we use original_ee here?

File: test/test_evaluator.cpp [hunk 6]
```diff
@@ -558,4 +558,101 @@ TEST_F(ExprEvalTest, Permute) {
...
+  out = permute(out, {1, 2, 0});
+  out = reshape(out, {IrBuilder::create<Val>(6), size(out, 2)});
+  fusion.addOutput(out);
+
+  at::Tensor in_tensor = at::rand({72}).cuda().as_strided({9, 6}, {8, 1});
+
+  ExpressionEvaluator evaluator;
+  evaluator.bind(in, in_tensor);
```
Comment (Priya2698) [inline test/test_evaluator.cpp:577]: There is a bind operation in the `evaluate` method, so is this needed here?

Comment (wujingyue) [reply test/test_evaluator.cpp:577]: Good catch! In fact, it's the other way around. `ViewOp::evaluate` shouldn't call `bind`. [Code here](https://github.com/NVIDIA/Fuser/blob/6751f27a8e128c427f6fd4aedd58732e37fb4013/csrc/expr_evaluator.cpp#L201) finds dependencies and evaluates them recursively.

Comment (wujingyue) [reply test/test_evaluator.cpp:577]: Done.

**Review by Priya2698 — APPROVED** [review]
LGTM.

**Pull request merged**